data_steward/test/unit_test/data_steward/cdr_cleaner/cleaning_rules/drop_participants_without_ppi_or_ehr_test.py

```diff
@@ -0,0 +1,28 @@
+# Python imports
+import copy
+import unittest
+
+# Third party imports
+
+# Project imports
+import common
+from constants import bq_utils as bq_consts
+from constants.cdr_cleaner import clean_cdr as clean_consts
+from cdr_cleaner.cleaning_rules import drop_rows_for_missing_persons
+from cdr_cleaner.cleaning_rules import drop_participants_without_ppi_or_ehr
+import resources
+
+
+class DropParticipantsWithoutPpiOrEhrTest(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        print('**************************************************************')
+        print(cls.__name__)
+        print('**************************************************************')
+
+    def test_get_queries(self):
+        results = drop_participants_without_ppi_or_ehr.get_queries('foo', 'bar')
+
+        self.assertEquals(len(results), 1 + len(common.CLINICAL_DATA_TABLES),
```

Comment:
The testing story looks bleak here, I didn't see any clear way of testing my code that wasn't effectively a change-detector test. This may be a larger discussion, but things I might hope to have in a hermetic local test:

1. SQL validation
2. Execution against an in-memory DB (with workarounds where needed)
3. (maybe - if I can't have 1/2) golden file SQL output tests

Reply:
I agree, this is a larger discussion and also on the desired local test features. If I understand 3 correctly, I think one problem with golden files is they may be sensitive to changes that are not functional. For example, I'm not sure we want solutions confined to be particular about whitespace, casing, or even whether a query uses a CTE vs subqueries (if both can do the job just as well).

Reply:
Yes, golden file testing has many problems. Setting aside whitespace changes (these would hopefully be formatted uniformly by your golden data test / generation), they do not actually do any semantic testing or validation of the SQL. This is only a last resort if 1,2 are too high effort; the reason I even mention the option is that I see a bunch of string-based testing in some of the cleaning rule units tests - these have the same deficiencies as golden file tests but with higher development overhead.

I am interested in exploring whether https://github.com/verilylifesciences/purplequery can work for hermetic fake BQ testing. It was recently created by Verily, and I can't fully vouch for it yet - but if it works as advertised it would be a really nice solution for testing.